### DISCUSSION TITLE
[BE] Provider & Validator

### DISCUSSION THREAD

**Discussion by Gyaak** [Kernel360/F2-BAGUNI#66] [메모장]
9/26(목)에 진행된 멘토링에서 `Provider`와 `Validator`에 대한 이야기가 있었습니다.
---
### Provider
- `service`와 `repository` 사이의 **중간다리** 역할
- 객체를 얻는 과정에서 발생하는 예외처리 등을 담당
```
@Component
@RequiredArgsConstructor
public class LinkProvider {

	private final LinkRepository linkRepository;

	public Link findById(Long linkId) throw ApiLinkException {

		return linkRepository.findById(linkId).orElseThrow(ApiLinkException::LINK_NOT_FOUND);
	}
}
```
- `service`에서는 `provider`를 통해 객체를 전달받으니 예외처리와 같은 부가적인 로직을 생각하지 않아도 됨
  -> 비즈니스 로직에 집중 가능
---
### Validator
- 엔티티에 대한 검증을 담당하는 클래스
- db에 이미 존재하는 엔티티인지(중복검사), 올바른 형식을 가지고 있는지 등을 확인
- @sangwonsheep 이전 프로젝트에서 Validator를 따로 둔 것으로 알고 있는데, 혹시 설명 중 부족한 점이 있다면 코멘트 부탁해요.

<br>

## [9/28 토요일까지 의견 부탁드립니다.]

<br>

**Comment by sangwonsheep** [comment]
이전 프로젝트에서 Provider + Validator 역할을 하는 것이 `Validator` 였습니다.
멘토링을 받고, Member 부분만 Provider로 이름만 변경하게 되었습니다.
Provider, Validator를 제대로 나누지 못했던 것 같습니다.

개인적으로, Provider만 사용하는 것을 선호합니다.
DB에서 값을 가져오는 코드를 Service에 작성하게 되면 코드가 상당히 지저분해지게 되어 Provider쪽에 작성하는 것이 맞다고 생각합니다.
중복 검사 부분의 경우, Service 코드에 작성을 하는 것을 선호합니다.
코드가 지저분 해질 수 있지만, 직관적으로 볼 수 있기 때문에 선호합니다.

중복 검사 부분을 Validator로 빼게 됐을 때 가지는 장점에 대해서 잘 모르겠습니다!
장점에 대해 알고계시면 알려주시면 감사하겠습니다!

**Comment by kimminkyeu** [comment]
@Gyaak @sangwonsheep 이 부분은 Service 내부에서 private 메소드로 구현하는 것으로 통일합니다. #82
